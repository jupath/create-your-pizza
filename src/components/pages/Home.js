import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import moment from 'moment';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { addOrder } from '../../actions/order';
import Pizza from '../layout/Pizza';
import CreateYourPizzaTable from '../layout/CreateYourPizzaTable';

const OrderTableWrapper = styled.div`

`;

export class Home extends Component {
  state = {
    sumPrice: 4,
    ingredients: [{ name: 'mushroom', isAdded: false, price: 0.9 }, { name: 'onion', isAdded: false, price: 0.8 }, { name: 'olive', isAdded: false, price: 1 }, { name: 'salami', isAdded: false, price: 1.5 }, { name: 'bacon', isAdded: false, price: 1.4 }],
  };

  handleAddIngredient = (name) => {
    const ingredients = this.state.ingredients.map((ingredient) => {
      if (ingredient.name === name && ingredient.isAdded !== true) {
        this.setState({ sumPrice: this.state.sumPrice + ingredient.price });
        return ({
          ...ingredient,
          isAdded: true,
        });
      }
      return ingredient;
    });
    this.setState({ ingredients });
  }

  handleRemoveIngredient = (name) => {
    const ingredients = this.state.ingredients.map((ingredient) => {
      if (ingredient.name === name && ingredient.isAdded === true) {
        this.setState({ sumPrice: this.state.sumPrice - ingredient.price });
        return ({
          ...ingredient,
          isAdded: false,
        });
      }
      return ingredient;
    });
    this.setState({ ingredients });
  }

  handleOrder = () => {
    const ingredients = this.state.ingredients
      .filter(ingredient => ingredient.isAdded === true)
      .map(ingredient => ingredient.name);

    const price = this.state.sumPrice.toFixed(2);

    const order = {
      time: moment().format(),
      ingredients,
      price,
    };

    this.props.addOrder(order);
    this.props.history.push('/order');
  }

  render() {
    return (
      <Fragment>
        <Pizza />
        <OrderTableWrapper>
          <CreateYourPizzaTable
            sumPrice={this.state.sumPrice}
            ingredients={this.state.ingredients}
            handleAddIngredient={this.handleAddIngredient}
            handleRemoveIngredient={this.handleRemoveIngredient}
            handleOrder={this.handleOrder}
          />
        </OrderTableWrapper>
      </Fragment>
    );
  }
}

Home.propTypes = {
  addOrder: PropTypes.func.isRequired,
};

const mapDispatchToProps = dispatch => ({
  addOrder: order => dispatch(addOrder(order)),
});

export default connect(null, mapDispatchToProps)(Home);
