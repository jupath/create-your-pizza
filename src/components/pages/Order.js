import React, { Component } from 'react';
import { connect } from 'react-redux';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import database from '../../firebase/firebase';
import OrderForm from '../layout/OrderForm';
import colors from '../../styles/colors';

const SuccessMessage = styled.h2`
  color: ${colors.green};
  margin-top: 100px;
  text-align: center;
`;

class Order extends Component {
  state = {
    sentOrder: false,
  }

  sendOrder = (shipping, order) => {
    const orderDetails = {
      shipping,
      order,
    };

    if (this.props.uid !== '') {
      database.ref(`users/${this.props.uid}/orders`)
        .push(orderDetails)
        .then(() => this.setState({ sentOrder: true }));
    } else {
      this.setState({ sentOrder: true });
    }
  }

  render() {
    return (
      <div>
        {
          this.state.sentOrder ?
            <SuccessMessage>Thanks, your order will arrive soon!</SuccessMessage>
          :
            <OrderForm sendOrder={this.sendOrder} />
        }
      </div>
    );
  }
}

Order.propTypes = {
  uid: PropTypes.string.isRequired,
};

const mapStateToProps = state => ({
  uid: state.auth.uid,
});

export default connect(mapStateToProps)(Order);
