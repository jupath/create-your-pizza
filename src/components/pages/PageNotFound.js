import React from 'react';
import styled from 'styled-components';

const NotFound = styled.h2`
  margin-top: 40px;
  text-align: center;
`;

export const PageNotFound = () => (
  <NotFound>404 error! Page Not Found!</NotFound>
);

export default PageNotFound;
