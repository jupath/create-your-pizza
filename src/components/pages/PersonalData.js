import React from 'react';

export const PersonalData = () => (
  <div>
    Hello
  </div>
);

export default PersonalData;
