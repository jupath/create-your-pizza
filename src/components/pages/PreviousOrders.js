import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import database from '../../firebase/firebase';
import PreviousOrdersItem from '../layout/PreviousOrdersItem';
import { Container, Title } from '../../styles/general';
import { Table, Thead, Tbody } from '../../styles/table';

export class PreviousOrders extends Component {
  state = {
    previousOrders: [],
  }

  componentDidMount() {
    this.getPreviousOrders();
  }

  getPreviousOrders = () => {
    database.ref(`users/${this.props.uid}/orders`)
      .once('value')
      .then((snapshot) => {
        const previousOrders = [];
        snapshot.forEach((childSnapshot) => {
          previousOrders.push({
            id: childSnapshot.key,
            ...childSnapshot.val(),
          });
        });
        this.setState({ previousOrders });
      });
  }

  render() {
    return (
      <Container>
        <Title>My previous orders</Title>
        <Table>
          <Thead>
            <tr><td>Date</td><td>Toppings</td><td>Price</td></tr>
          </Thead>
          <Tbody>
            {
              this.state.previousOrders.length > 0 ?
              this.state.previousOrders.map(previousOrder => <PreviousOrdersItem key={previousOrder.id} previousOrder={previousOrder} />) :
              <tr><td colSpan="3">You have no previous orders</td></tr>
            }
          </Tbody>
        </Table>
      </Container>
    );
  }
}

PreviousOrders.propTypes = {
  uid: PropTypes.string.isRequired,
};

const mapStateToProps = state => ({
  uid: state.auth.uid,
});

export default connect(mapStateToProps)(PreviousOrders);
