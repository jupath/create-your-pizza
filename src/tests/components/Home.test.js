import React from 'react';
import { shallow } from 'enzyme';
import moment from 'moment';
import { Home } from '../../components/pages/Home';

let addOrder, history, wrapper;

beforeEach(() => {
  addOrder = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(<Home addOrder={addOrder} history={history} />);
});

test('should render Home correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should add ingredient', () => {
  wrapper.instance().handleAddIngredient('mushroom');
  const addedIngredient = wrapper.state('ingredients').filter(ingredient => ingredient.name === 'mushroom');
  expect(addedIngredient[0].isAdded).toBe(true);
  expect(wrapper.state('sumPrice')).toBe(4.9);
});

test('should remove ingredient', () => {
  wrapper.instance().handleAddIngredient('onion');
  wrapper.instance().handleRemoveIngredient('onion');
  const removedIngredient = wrapper.state('ingredients').filter(ingredient => ingredient.name === 'onion');
  expect(removedIngredient[0].isAdded).toBe(false);
  expect(wrapper.state('sumPrice')).toBe(4);
});

test('should call addOrder', () => {
  const ingredients = ['mushroom'];
  const price = '4.90';
  const order = {
    time: moment().format(),
    ingredients,
    price,
  };
  wrapper.instance().handleAddIngredient('mushroom');
  wrapper.instance().handleOrder();
  expect(addOrder).toHaveBeenLastCalledWith(order);
  expect(history.push).toHaveBeenLastCalledWith('/order');
});
