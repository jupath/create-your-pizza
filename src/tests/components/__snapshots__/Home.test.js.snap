// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Home correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Home
    addOrder={[Function]}
    history={
        Object {
            "push": [Function],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <Pizza />,
        <styled.div>
          <CreateYourPizzaTable
                    handleAddIngredient={[Function]}
                    handleOrder={[Function]}
                    handleRemoveIngredient={[Function]}
                    ingredients={
                              Array [
                                        Object {
                                          "isAdded": false,
                                          "name": "mushroom",
                                          "price": 0.9,
                                        },
                                        Object {
                                          "isAdded": false,
                                          "name": "onion",
                                          "price": 0.8,
                                        },
                                        Object {
                                          "isAdded": false,
                                          "name": "olive",
                                          "price": 1,
                                        },
                                        Object {
                                          "isAdded": false,
                                          "name": "salami",
                                          "price": 1.5,
                                        },
                                        Object {
                                          "isAdded": false,
                                          "name": "bacon",
                                          "price": 1.4,
                                        },
                                      ]
                    }
                    sumPrice={4}
          />
</styled.div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <CreateYourPizzaTable
            handleAddIngredient={[Function]}
            handleOrder={[Function]}
            handleRemoveIngredient={[Function]}
            ingredients={
                        Array [
                                    Object {
                                      "isAdded": false,
                                      "name": "mushroom",
                                      "price": 0.9,
                                    },
                                    Object {
                                      "isAdded": false,
                                      "name": "onion",
                                      "price": 0.8,
                                    },
                                    Object {
                                      "isAdded": false,
                                      "name": "olive",
                                      "price": 1,
                                    },
                                    Object {
                                      "isAdded": false,
                                      "name": "salami",
                                      "price": 1.5,
                                    },
                                    Object {
                                      "isAdded": false,
                                      "name": "bacon",
                                      "price": 1.4,
                                    },
                                  ]
            }
            sumPrice={4}
/>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "handleAddIngredient": [Function],
            "handleOrder": [Function],
            "handleRemoveIngredient": [Function],
            "ingredients": Array [
              Object {
                "isAdded": false,
                "name": "mushroom",
                "price": 0.9,
              },
              Object {
                "isAdded": false,
                "name": "onion",
                "price": 0.8,
              },
              Object {
                "isAdded": false,
                "name": "olive",
                "price": 1,
              },
              Object {
                "isAdded": false,
                "name": "salami",
                "price": 1.5,
              },
              Object {
                "isAdded": false,
                "name": "bacon",
                "price": 1.4,
              },
            ],
            "sumPrice": 4,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Pizza />,
          <styled.div>
            <CreateYourPizzaTable
                        handleAddIngredient={[Function]}
                        handleOrder={[Function]}
                        handleRemoveIngredient={[Function]}
                        ingredients={
                                    Array [
                                                Object {
                                                  "isAdded": false,
                                                  "name": "mushroom",
                                                  "price": 0.9,
                                                },
                                                Object {
                                                  "isAdded": false,
                                                  "name": "onion",
                                                  "price": 0.8,
                                                },
                                                Object {
                                                  "isAdded": false,
                                                  "name": "olive",
                                                  "price": 1,
                                                },
                                                Object {
                                                  "isAdded": false,
                                                  "name": "salami",
                                                  "price": 1.5,
                                                },
                                                Object {
                                                  "isAdded": false,
                                                  "name": "bacon",
                                                  "price": 1.4,
                                                },
                                              ]
                        }
                        sumPrice={4}
            />
</styled.div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <CreateYourPizzaTable
              handleAddIngredient={[Function]}
              handleOrder={[Function]}
              handleRemoveIngredient={[Function]}
              ingredients={
                            Array [
                                          Object {
                                            "isAdded": false,
                                            "name": "mushroom",
                                            "price": 0.9,
                                          },
                                          Object {
                                            "isAdded": false,
                                            "name": "onion",
                                            "price": 0.8,
                                          },
                                          Object {
                                            "isAdded": false,
                                            "name": "olive",
                                            "price": 1,
                                          },
                                          Object {
                                            "isAdded": false,
                                            "name": "salami",
                                            "price": 1.5,
                                          },
                                          Object {
                                            "isAdded": false,
                                            "name": "bacon",
                                            "price": 1.4,
                                          },
                                        ]
              }
              sumPrice={4}
/>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "handleAddIngredient": [Function],
              "handleOrder": [Function],
              "handleRemoveIngredient": [Function],
              "ingredients": Array [
                Object {
                  "isAdded": false,
                  "name": "mushroom",
                  "price": 0.9,
                },
                Object {
                  "isAdded": false,
                  "name": "onion",
                  "price": 0.8,
                },
                Object {
                  "isAdded": false,
                  "name": "olive",
                  "price": 1,
                },
                Object {
                  "isAdded": false,
                  "name": "salami",
                  "price": 1.5,
                },
                Object {
                  "isAdded": false,
                  "name": "bacon",
                  "price": 1.4,
                },
              ],
              "sumPrice": 4,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
